### PR TITLE
Polio 670 destruction update

### PR DESCRIPTION
**Pull request by quang-le**
There was only 1 destruction of vaccines per rounds, but we need many

## Self proof reading checklist

- [X] Did I use eslint and black formatters
- [X] Is my code clear enough and well documented
- [X] Are my typescript files well typed
- [X] Did I add translations
- [X] My migrations file are included

## Changes

- Created a `Destruction` model
- Updated `RoundSerializer` to use the new model
- Created `DestructionForm` and `DestructionsForm` to handle adding and removing destructions
- Updated yup validation to reflect changes
- fixed some translation typos
- removed validator error on number fields with value 0
- Add fix for `TextArea` infinite loop bug

## How to test

Go to campaigns, open one, go Vaccines tab and try adding and removing destructions

## Print screen / video

https://user-images.githubusercontent.com/38907762/200021794-18db2189-fd96-4a3d-a847-603ab2d2f880.mov

## Notes

There's room for improvement as most code (both front and back) is pretty much copy pasted from Shipments

### CODE REVIEW

File: plugins/polio/js/src/constants/translations/en.json [hunk 1]
```diff
@@ -1,4 +1,5 @@
 {
+    "Date of the destruction report": "Date of the destruction report",
```
Comment (hakifran) [inline plugins/polio/js/src/constants/translations/en.json:2]: This shouldn't be there.

File: plugins/polio/js/src/constants/translations/fr.json [hunk 2]
```diff
@@ -103,6 +103,7 @@
     "iaso.polio.import_line_list": "Importer Line List",
     "iaso.polio.import_line_list.successMessage": "Import réussi pour {amount} campagne(s)",
     "iaso.polio.label.add": "Ajouter",
+    "iaso.polio.label.addDestruction": "Ajouter destruction",
```
Comment (hakifran) [inline plugins/polio/js/src/constants/translations/fr.json]: ce ne serait pas mieux si on disait "Ajouter une destruction"?

File: plugins/polio/js/src/constants/translations/fr.json [hunk 3]
```diff
@@ -286,6 +289,7 @@
     "iaso.polio.label.rejectBudgetForCampaign": "Rejeter le budget pour la campagne: {campaign}",
     "iaso.polio.label.rejected": "Rejeté",
     "iaso.polio.label.Religious_leader": "Leader religieux",
+    "iaso.polio.label.removeLastDestruction": "Effacer dernière destruction",
```
Comment (hakifran) [inline plugins/polio/js/src/constants/translations/fr.json]: "Effacer la dernière destruction"

File: plugins/polio/js/src/forms/DestructionsForm.tsx [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+}) => {
+    const { formatMessage } = useSafeIntl();
+    const { setFieldValue } = useFormikContext();
+    const { destructions = [] } = round ?? {};
+    const [enableRemoveButton, setEnableRemoveButton] =
+        useState<boolean>(false);
+
+    const lastIndex = destructions.length >= 1 && destructions.length - 1;
+
+    const handleAddShipment = () => {
```
Comment (hakifran) [inline plugins/polio/js/src/forms/DestructionsForm.tsx]: This method should be renamed to "handleAddDestruction"

File: plugins/polio/js/src/forms/DestructionsForm.tsx [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+        const newShipments = [...destructions, {}];
+        setFieldValue(`${accessor}.destructions`, newShipments);
+    };
+
+    const handleRemoveLastShipment = useCallback(() => {
```
Comment (hakifran) [inline plugins/polio/js/src/forms/DestructionsForm.tsx]: This method too "handleRemoveLastDestruction"

File: plugins/polio/js/src/forms/DestructionsForm.tsx [hunk 6]
```diff
@@ -0,0 +1,108 @@
...
+        const newShipments = [...destructions];
```
Comment (hakifran) [inline plugins/polio/js/src/forms/DestructionsForm.tsx]: Here too newDestructions

File: plugins/polio/js/src/forms/VaccineManagementForm.tsx [hunk 7]
```diff
@@ -192,11 +193,16 @@ export const VaccineManangementForm: FunctionComponent<Props> = () => {
                         </Typography>
                     </Box>
                 </Grid>
-                <Grid container direction="row" spacing={2} item xs={12}>
-                    {/* Not sure yet if this should not be a map */}
+                <DestructionsForm
+                    accessor={accessor}
+                    round={rounds[roundIndex]}
+                />
+                {/* <Grid container direction="row" spacing={2} item xs={12}>
```
Comment (hakifran) [inline plugins/polio/js/src/forms/VaccineManagementForm.tsx]: This commented code be removed to make the code more clean. thank you

**Review by hakifran — COMMENTED** [review]
Apart from some translation and methods name to change, it seems working for me and the remaining code seems to me correct. thank you @quang-le

File: plugins/polio/js/src/forms/DestructionsForm.tsx (hunk 4 above)
Comment (quang-le) [reply plugins/polio/js/src/forms/DestructionsForm.tsx]: Me copy/pasting my own code?😅

**Review by beygorghor — CHANGES_REQUESTED** [review]
Screenshot 2022-11-07 at 12 07 32
Screenshot 2022-11-07 at 12 07 09

I have an error on textArea while i shouldn't
Screenshot 2022-11-07 at 12 07 22
Also adding some more space before the divider if possible

**Comment by quang-le** [comment]
> 

It seems validation for Shipments broke somewhere on the way

**Comment by codeKameleon** [comment]
@quang-le I think it would be more clear to the user if we add required asterisk to the inputs that need to be filled before saving

File: plugins/polio/js/src/forms/DestructionsForm.tsx [hunk 8]
```diff
@@ -0,0 +1,108 @@
...
+        useState<boolean>(false);
+
+    const lastIndex = destructions.length >= 1 && destructions.length - 1;
+
+    const handleAddDestruction = () => {
+        const newDestructions = [...destructions, {}];
+        setFieldValue(`${accessor}.destructions`, newDestructions);
+    };
+
+    const handleRemoveLastDestruction = useCallback(() => {
```
Comment (codeKameleon) [inline plugins/polio/js/src/forms/DestructionsForm.tsx:36]: @quang-le Is there a use case where the user would like to delete a older destruction ? For the moment the feature only allow to remove the last one

**Review by codeKameleon — CHANGES_REQUESTED** [review]

File: plugins/polio/js/src/forms/DestructionsForm.tsx (hunk 8 above)
Comment (quang-le) [reply plugins/polio/js/src/forms/DestructionsForm.tsx:36]: Not really. They might want to edit it, but once the vaccines are destroyed, they are destroyed. Same with Shipments

**Comment by quang-le** [comment]
> @quang-le I think it would be more clear to the user if we add required asterisk to the inputs that need to be filled before saving

The thing is that it's conditional: all destruction inputs should be filled for a line (except comment) only if one has been filled. This as per client's request. Same with Shipments

**Comment by quang-le** [comment]
> > 
> 
> It seems validation for Shipments broke somewhere on the way

It's not that, it's just that for new campaigns, the whole form will error when there's no obr_name. It works fine otherwise

**Comment by codeKameleon** [comment]
> > @quang-le I think it would be more clear to the user if we add required asterisk to the inputs that need to be filled before saving
> 
> The thing is that it's conditional: all destruction inputs should be filled for a line (except comment) only if one has been filled. This as per client's request. Same with Shipments

Ok i see, so maybe in a future PR, show an info message in the Destruction section if the user put a value on one of the inputs, we could show "These fiellds need to be filled ..."

**Comment by olethanh** [comment]
EDIT : Apparently I cannot read. You already did the AuditDestructionSerializer, you just need to add it to the AuditRoundSerializer

You need to add the new Models to the Audit so the change modifications get kept.

See following PR for example on how to do this
https://github.com/BLSQ/iaso/pull/60/files

But basically you need to ad an AuditSerializer and then reference it in the

**Review by hakifran — APPROVED** [review]
Seems good to me. thank you @quang-le

**Pull request merged**